src/vmspawn/vmspawn.c

```diff
@@ -474,20 +488,124 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
+
+        if (ret_runtime_directory_property)
+                *ret_runtime_directory_property = TAKE_PTR(runtime_directory_property);
+
+        if (ret_tempdir || ret_runtime_directory_property)
+                TAKE_PTR(tempdir);
+
+        return 0;
+}
+
+static int start_tpm(sd_bus *bus, const char *scope, const char *our_runtime_dir, const char *tpm, const char **ret_state_tempdir) {
+        _cleanup_(rm_rf_physical_and_freep) char *state_dir = NULL;
+        _cleanup_strv_free_ char **cmdline = NULL;
+        char **extra_properties;
+        _cleanup_free_ char *sock_path = NULL, *scope_prefix = NULL, *unit_name_prefix = NULL, *state_runtime_dir = NULL;
+        int r;
+
+        assert(bus);
+        assert(scope);
+        assert(tpm);
+        assert(our_runtime_dir);
+        assert(ret_state_tempdir);
+
+        r = create_runtime_tempdir(our_runtime_dir, "tpm-XXXXXX", &state_dir, &state_runtime_dir);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create runtime tempdir: %m");
+
+        extra_properties = STRV_MAKE(state_runtime_dir);
+
+        sock_path = path_join(state_dir, "sock");
+        if (!sock_path)
+                return log_oom();
+
+        cmdline = strv_new(tpm, "socket", "--tpm2", "--tpmstate");
+        if (!cmdline)
+                return log_oom();
+
+        r = strv_extendf(&cmdline, "dir=%s", state_dir);
+        if (r < 0)
+                return log_oom();
+
+        r = strv_extend_strv(&cmdline, STRV_MAKE("--ctrl", "type=unixio,fd=3"), /* filter_duplicates= */ false);
+        if (r < 0)
+                return log_oom();
+
+        r = unit_name_to_prefix(scope, &scope_prefix);
+        if (r < 0)
+                return log_error_errno(r, "Failed to strip .scope suffix from scope: %m");
+
+        unit_name_prefix = strjoin(scope_prefix, "-tpm");
+        if (!unit_name_prefix)
+                return log_oom();
+
+        r = attach_command_to_socket_in_scope(bus, scope, unit_name_prefix, sock_path, SOCK_STREAM, cmdline, NULL, extra_properties);
+        if (r < 0)
+                return r;
+
+        *ret_state_tempdir = TAKE_PTR(state_dir);
+
+        return 0;
+}
+
 static int run_virtual_machine(void) {
         _cleanup_(ovmf_config_freep) OvmfConfig *ovmf_config = NULL;
+        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
+
+        _cleanup_close_ int vsock_fd = -EBADF;
+        _cleanup_free_ char *machine = NULL, *qemu_binary = NULL, *mem = NULL, *trans_scope = NULL,
+                            *our_runtime_dir = NULL;
+
         _cleanup_strv_free_ char **cmdline = NULL;
-        _cleanup_free_ char *machine = NULL, *qemu_binary = NULL, *mem = NULL;
         int r;
-        _cleanup_close_ int vsock_fd = -EBADF;
+
+        if (getuid() == 0)
+                r = sd_bus_open_system(&bus);
+        else
+                r = sd_bus_open_user(&bus);
+        if (r < 0)
+                return log_error_errno(r, "Failed to connect to systemd bus: %m");
```

Comment:
sd_bus_default()?

Reply:
I believe that this is actually the correct way of doing this, without it vmspawn cannot open a bus connection when invoked as root.